packages/ui/src/components/controls/NumberInput.tsx

```diff
@@ -0,0 +1,126 @@
+import clsx from 'clsx';
+import type {JSX} from 'preact';
+import {useCallback, useEffect, useRef, useState} from 'preact/hooks';
+import {useDocumentEvent} from '../../hooks';
+import {MouseButton, clamp} from '../../utils';
+import styles from './Controls.module.scss';
+
+type NumberInputProps = Omit<
+  JSX.HTMLAttributes<HTMLInputElement>,
+  'min' | 'max' | 'step' | 'value' | 'onChange' | 'onChangeCapture'
+> & {
+  min?: number;
+  max?: number;
+  step?: number;
+  value: number;
+  onChange?: (value: number) => void;
+  onChangeCapture?: (value: number) => void;
+};
+
+export function NumberInput({
+  min = -Infinity,
+  max = Infinity,
+  step = 1,
+  value = 0,
+  onChange,
+  onChangeCapture,
+  ...props
+}: NumberInputProps) {
+  const handleChange = onChangeCapture ? onChange : undefined;
+  const handleChangeCapture = onChangeCapture ?? onChange;
+
+  const inputRef = useRef<HTMLInputElement>();
+  const [currentValue, setCurrentValue] = useState(value);
+  const [active, setActive] = useState(false);
+  const [startX, setStartX] = useState(0);
+
+  useEffect(() => {
+    setCurrentValue(value);
+  }, [value]);
+
+  const handleDrag = useCallback(
+    (event: MouseEvent) => {
+      setCurrentValue(value => {
+        // ignore very large jumps in movement, as they often result from
+        // mouse acceleration issues
+        if (Math.abs(event.movementX) > 100) {
+          return value;
+        }
+
+        const newValue = clamp(min, max, value + event.movementX * step);
+        handleChange?.(newValue);
+        return newValue;
+      });
+    },
+    [min, max, step],
+  );
+
+  useDocumentEvent('pointerlockchange', () => {
+    if (document.pointerLockElement === inputRef.current) {
+      document.addEventListener('pointermove', handleDrag);
+    } else {
+      document.removeEventListener('pointermove', handleDrag);
+    }
+  });
```

Comment:
This captures `handleDrag` and won't update it even when `min`, `max`, or `step` change.

Reply:
No, never mind. I thought the second argument was wrapped in `useCallback` since that's how `useDocumentEvent` is usually used but you pass a raw function so the whole `pointerlockchange` will be resubscribed on each re-render. Is this something you do on purpose? Because I've noticed you removed `useCallback` from the timeline document event as well.

Reply:
In that case I did it on purpose for exactly that reason but if you have a more elegant solution please let me know! To be honest I don't exactly remember what I was thinking when working on the timeline, I'd hope there was a reason for that as well.. :)

Reply:
I'm not sure about elegant but usually you'd use `useCallback` to memoize the function and pass that to the hook:
```suggestion
  useDocumentEvent(
    'pointerlockchange',
    useCallback(() => {
      if (document.pointerLockElement === inputRef.current) {
        document.addEventListener('pointermove', handleDrag);
      } else {
        document.removeEventListener('pointermove', handleDrag);
      }
    }, [handleDrag]),
  );
```
This will only re-register the handler when `handleDrag` or its dependencies change.
I'm planning on improving `useDocumentEvent` so that it doesn't rely on external memoization but right now this would be the way to go.

Reply:
Alright, sounds good!